cpp/oneapi/dal/backend/primitives/placement/search_sorted_dpc.cpp

```diff
@@ -0,0 +1,126 @@
+/*******************************************************************************
+* Copyright 2023 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <algorithm>
+#include <functional>
+
+#include "oneapi/dal/detail/profiler.hpp"
+
+#include "oneapi/dal/backend/primitives/common.hpp"
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+
+#include "oneapi/dal/backend/primitives/placement/search_sorted.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+template <search_alignment alignment, typename Type>
+struct comparator {};
+
+template <typename Type>
+struct comparator<search_alignment::left, Type> {
+    constexpr static auto value = std::less<Type>{};
+};
+
+template <typename Type>
+struct comparator<search_alignment::right, Type> {
+    constexpr static auto value = std::less_equal<Type>{};
+};
+
+template <search_alignment alignment, typename Type>
+constexpr auto comparator_v = comparator<alignment, Type>::value;
+
+template <bool clip, typename Index>
+inline Index clip_place(std::int64_t count, Index result) {
+    if constexpr (clip) {
+        return std::max(static_cast<Index>(0), std::min<Index>(result, count - 1));
+    }
+    else {
+        return result;
+    }
+}
+
+// Reference numpy implementation:
+// https://github.com/numpy/numpy/blob/main/numpy/core/src/npysort/binsearch.cpp#L61
```

Comment:
That's great there is a link to the reference implementation (not joking).
But it can be moved from #L61 in the future. Can you please provide a more stable description?

Reply:
Done. Switched to a release-specific branch that should not change.